README.rst
.. image:: https://travis-ci.org/jacopofar/flask-typed-mounter.svg?branch=master
    :target: https://travis-ci.org/jacopofar/flask-typed-mounter
    :alt: Travis CI badge

Flask typed mounter
###################


Exposes a plain Python function as an HTTP endpoint using Flask, performs type checking between HTTP request and Python 3.6 type hints to give the client a clear error before calling the function. A GET request retrieves the documentation and type hints for that function

Work in progress, don't hesitate forking/asking/suggesting


See `example.py` for a complete example, in short:

.. code-block:: python
    @tm.attach_endpoint('/mul', methods=['POST'], auto_document=True)
    def multiplier(val1: int, val2: int = 5):
        return val1 * val2


this will expose your multiplier function at `/mul`, with the result that a POST of a JSON will be validated against type hints and give a JSON answer or nice error message if not matching. A GET for that endpoint will show the rendered pydoc string and the type hints.

The endpoint can be called by sending a JSON or a form, the former allows for more structured requests and the second allows to send files.

Type checking (based on Python 3.6 typing):

.. image:: error_check.png

Documentation:

.. image:: doc_screenshot.png

Correct call:

.. image:: no_errors.png


A customized doc template can be passed as `doc_html_template` when instantiating the server and documentation endpoint can be disabled on single functions by passing `auto_document=False`.

You can also accept files (which by default is not allowed) by passing the `accept_files` parameter set to `True`. Additionally, you can limit the file extensions by passing the `allowed_extensions` list; if this parameter is omitted, any extension is accepted.

By default there is **no limit on the size of the files** which can be uploaded. You can set it up in Flask by setting `app.config['MAX_CONTENT_LENGTH']` to the maximum number of bytes.

Files will be saved in a temporary directory and be given to the function as a `pathlib.Path` object, if and only if that's the type annotation used by the function. Files are deleted after the function has been invoked, but the function itself can of course copy the contend during its execution.

CHANGELOG.md
# Changelog
All notable changes to Flask typed mounter will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.2.1] 2017-09-30
- Accept form and allow file upload

## [0.2.0] 2017-09-29
- Show exceptions in HTTP response and log

## [0.1.2] 2017-09-29
- Nicer documentation template, thanks to [NikolaJankovic](https://github.com/NikolaJankovic)
- Add docutils to the setuptools dependencies

## [0.1] 2017-09-26
First release, attach a function to a POST endpoint, show documentation using GET

tests/test_api.py
import json
import unittest

from flask import Flask

from flask_typed_mounter import TypedMounter


class AppTest(unittest.TestCase):
    def setUp(self):
        app = Flask(__name__)
        self.tm = TypedMounter(app)
        self.app = app.test_client()

        @self.tm.attach_endpoint('/mul', methods=['POST'], auto_document=False)
        def multiplier(val1: int, val2: int = 5):
            return val1 * val2

        @self.tm.attach_endpoint('/mul_two', methods=['POST'])
        def multiplier_with_doc(val1: int, val2: int = 5):
            """
            SPECIAL_STRING_FROM_DOCUMENTATION
            """
            return val1 * val2

        @self.tm.attach_endpoint('/concat', methods=['POST'])
        def concat(a: str, b: str = 'yeah'):
            return ''.join([a, b])

    def test_correct_call(self):
        response = self.app.post("/mul",
                                 data=json.dumps({'val1': 4}),
                                 content_type='application/json')
        self.assertEqual(response.status_code, 200)

    def test_wrong_call(self):
        response = self.app.post("/mul",
                                 data=json.dumps({'val1': 'not an integer!'}),
                                 content_type='application/json')
        self.assertEqual(response.status_code, 400)

    def test_doc_generation(self):
        response = self.app.get("/mul_two")
        self.assertTrue('SPECIAL_STRING_FROM_DOCUMENTATION'.encode('utf-8') in response.data)
        self.assertEqual(response.status_code, 200)

    def test_form_call(self):
        response = self.app.post("/concat",
                                 data=dict(
                                     a='dog',
                                     b='cat'
                                 ))
        self.assertEqual(response.data.decode(), 'dogcat')

    def test_form_call_conversion(self):
        response = self.app.post("/concat",
                                 data=dict(
                                     a=12,
                                     b='monkeys'
                                 ))
        self.assertEqual(response.data.decode(), '12monkeys')

    def test_wrong_form_call(self):
        response = self.app.post("/concat",
                                 data=dict(
                                     b='monkeys'
                                 ))
        self.assertEqual(response.status_code, 400)


if __name__ == "__main__":
    unittest.main()


tests/test_raises.py
import unittest

from flask_typed_mounter import TypedMounter

class RaisesTest(unittest.TestCase):
    def setUp(self):
        self.tm = TypedMounter()

    def test_null_app_error(self):
        with self.assertRaises(ValueError):
            @self.tm.attach_endpoint('/mul', methods=['GET', 'POST'])
            def multiplier(val1: int, val2: int = 5):
                return val1 * val2

    def test_only_post_is_accepted(self):
        with self.assertRaises(ValueError):
            @self.tm.attach_endpoint('/mul_three', methods=['POST', 'PUT'])
            def foo(val1: int, val2: int = 5):
                return val1 * val2


if __name__ == '__main__':
    unittest.main()


example.py
from pathlib import Path
from typing import List

from flask import Flask

from flask_typed_mounter import TypedMounter


app = Flask(__name__)
tm = TypedMounter(app)


@tm.attach_endpoint('/mul', methods=['POST'], auto_document=True)
def multiplier(val1: int, names:List[str], val2: int = 5) -> str:
    """
    A function which retrieves a value from the given *list* with a given **index**
    and returns it a given number of times 😲

    :param names: list of names
    :param val1: a number
    :param val2: another number, defaults to 5
    :return: the product of the two values
    """
    if val2 ==7:
        print(23 / 0)
    return names[val1] * val2


@tm.attach_endpoint('/count_file_lines', methods=['POST'], auto_document=True, accept_files=True)
def multiplier(p: Path) -> int:
    """
    Count the  lines in a file
    :param p: the Path of the file
    :return: the number of lines
    """
    tot = 0
    for _ in open(p):
        tot += 1
    return tot


checker/app.py
import json
from functools import wraps

from flask import Flask, Response, request

from runtime_typecheck import check_args, DetailedTypeError

app = Flask(__name__)


def attach_endpoint(rule, **options):
    def actual_decorator(fun):
        # TODO parse options dictionary to retrieve options that are not among werkzeug ones
        # see http://werkzeug.pocoo.org/docs/0.12/routing/#werkzeug.routing.Rule
        # Example: allow automatic documentation when one calls GET or use GET parameters to call the function
        @wraps(fun)
        @app.route(rule, **options)
        def service():
            try:
                with_type_checking = check_args(fun)
                js = json.dumps(with_type_checking(**request.get_json()))
                resp = Response(js, status=200, mimetype='application/json')
                return resp

            except DetailedTypeError as dte:
                return Response(json.dumps({
                    "error": "invalid types",
                    "details": [{
                        'parameter': issue.name,
                        'expected': repr(issue.expected_type),
                        'found': issue.value,
                        'missing': issue.missing_parameter,
                        'message': issue.generic_message
                    } for issue in dte.issues]
                }), status=400, mimetype='application/json')

        return service
    return actual_decorator


@attach_endpoint('/mul', methods=['GET', 'POST'])
def multiplier(val1: int, val2: int = 5):
    return val1 * val2


flask_typed_mounter/typed_mounter.py
from functools import wraps
import inspect
import json
from pathlib import Path
from shutil import rmtree
import sys
import tempfile
from textwrap import dedent
import traceback

from docutils.core import publish_parts
from flask import Response, request
from jinja2 import Environment
from werkzeug.utils import secure_filename

from runtime_typecheck import check_args, DetailedTypeError


class TypedMounter(object):
    DEFAULT_DOC_HTML = '''
        <html>
        <head>
        <title>{{ function_name }}</title>
        <style>
        body {
            background-color: aliceblue;
            font-family: sans-serif;
        }
        table {
            width: 100%;
            border-collapse: collapse;
        }
        th, td {
            text-align: left;
            padding: 8px;
            border: 1px solid lightgray;
            color: rgba(89, 96, 105, 1.0);
        }
        th {
            color: rgba(36, 41, 46, 1.0);
            background: rgba(209, 230, 254, 0.7);
        }
        .function {
            padding: 10px;
            background-color: rgba(95, 183, 96, 1.0);
            border-radius: 5px;
            color: white;
            display: inline-block;
        }
        </style>
        </head>
        <body>
        <h1 class="function">{{ function_name }}</h1>
        <div class="doc">{{ doc_html|safe }}</div>
        <h4>Type Hints:</h4>
        <table class="type-hints-{{ function_name }}">
          <tr>
            <th>Parameter</th>
            <th>Type</th>
            <th>Default value</th>
          </tr>
            {% for p in parameters %}
                <tr>
                    <td>{{ p.name }}</td>
                    <td>{{ p.type }}</td>
                    <td>{{ p.default_value }}</td>
                </tr>
            {% endfor %}
        </table>
        </body>
        </html>
    '''

    def __init__(self, app=None, doc_html_template=DEFAULT_DOC_HTML):
        self._app = app
        self.doc_html = doc_html_template

    def init_app(self, app):
        self._app = app

    def attach_endpoint(self, rule, **options):
        if self._app is None:
            raise ValueError("App not initiated.")
        else:
            if options.get('methods', ['POST']) != ['POST']:
                raise ValueError(f'currently the function can be mounted only to a POST verb, was called with methods={options["methods"]}')

            def actual_decorator(fun):
                # keys used for the wrapper itself and not to be passed to Flask
                own_keys = ['auto_document', 'accept_files', 'allowed_extensions']
                if options.get('auto_document', True):
                    document_options = options.copy()
                    document_options['methods'] = ['GET']
                    for key in own_keys:
                        document_options.pop(key, None)

                    @self._app.route(rule, **document_options, endpoint=f'doc_{rule}')
                    def document():
                        doc_html = publish_parts(dedent(fun.__doc__), writer_name='html')['html_body']
                        parameter_descriptions = []
                        for name, attributes in inspect.signature(fun).parameters.items():
                            parameter_descriptions.append({
                                'name': name,
                                'type': repr(attributes.annotation),
                                'default_value': '' if attributes.default == inspect._empty else attributes.default
                            })
                        return Environment(autoescape=True)\
                            .from_string(dedent(self.doc_html))\
                            .render(doc_html=doc_html, parameters=parameter_descriptions, function_name=fun.__name__)


                # see http://werkzeug.pocoo.org/docs/0.12/routing/#werkzeug.routing.Rule
                # Example: allow automatic documentation when one calls GET or use GET
                # parameters to call the function.

                api_options = options.copy()
                api_options['methods'] = ['POST']
                for key in own_keys:
                    api_options.pop(key, None)

                @wraps(fun)
                @self._app.route(rule, endpoint=f'api_{rule}', **api_options)
                def service():

                    with_type_checking = check_args(fun)
                    try:
                        # JSON POST, this is what most of us want
                        if request.get_json() is not None:
                            js = json.dumps(with_type_checking(**request.get_json()))
                            resp = Response(js, status=200, mimetype='application/json')
                            return resp

                        # forms: no types, nor structure
                        # A simple dictionary where all values are strings, not very useful but still provided
                        # Since the input is flat, the desired output is likely plain text too
                        if len(request.values) + len(request.files) > 0:
                            form_params = {k: v for k, v in request.values.items()}
                            # now add the files, if any, to the argument dictionary
                            # presenting them as pathlib.Path instances
                            dir_for_request = None
                            if options.get('accept_files', False):
                                allowed_extensions = [ext.lower() for ext in options.get('allowed_extensions', [])]
                                for arg_name, file in request.files.items():
                                    if len(allowed_extensions) > 0:
                                        if file.filename.rsplit('.', 1)[1].lower() not in allowed_extensions:
                                            continue
                                    if dir_for_request is None:
                                        dir_for_request = Path(tempfile.mkdtemp())
                                        print('TMP directory:', dir_for_request)
                                    path = dir_for_request / secure_filename(file.filename)
                                    file.save(str(path))
                                    form_params[arg_name] = path
                            txt_response = str(with_type_checking(**form_params))
                            resp = Response(txt_response, status=200, mimetype='text/plain')
                            if dir_for_request is not None:
                                rmtree(str(dir_for_request))
                            return resp
                        return Response(f'Unknown request type. Mimetype was {request.mimetype}', status=400)
                    except DetailedTypeError as dte:
                        return Response(json.dumps({
                            "error": "invalid types",
                            "details": [{
                                'parameter': issue.name,
                                'expected': repr(issue.expected_type),
                                'found': issue.value,
                                'missing': issue.missing_parameter,
                                'message': issue.generic_message
                            } for issue in dte.issues]
                        }), status=400, mimetype='application/json')

                    except:
                        exc = sys.exc_info()
                        traceback.print_exc(file=sys.stdout)
                        return Response(str(f'Error in function {fun.__name__}\n {exc[0]}: {exc[1]}'), status=400)

                return service
        return actual_decorator
